### PR TITLE
Add `undo_last_commit` command

### PR DESCRIPTION
**Pull request by remique**
Hey again! I started working on #758 but since it’s not finished yet, I am opening Draft Pull Request. It works, but it’s not finished. Am I going in the right direction with this?
* It does not refresh main tab yet (sometimes). You have to switch between tabs to see the change 
* Should there be a popup or some kind of feedback?
* Should `undo_last_commit()` in `utils.rs` be more general function, for future use?

### CODE REVIEW

File: asyncgit/src/sync/utils.rs [hunk 1]
```diff
@@ -149,6 +149,21 @@ pub fn stage_add_all(repo_path: &str, pattern: &str) -> Result<()> {
     Ok(())
 }
 
+/// Undo last commit in repo
+pub fn undo_last_commit(repo_path: &str) -> Result<()> {
```
Comment (extrawurst) [inline asyncgit/src/sync/utils.rs:153]: please add a unittest to prove that it does what we want and that it fails as expected for example in an empty repo

**Comment by extrawurst** [comment]
@remique welcome back!🥳

> It does not refresh main tab yet (sometimes). You have to switch between tabs to see the change

you need to send a explicit update request into the queue (check other such functions)

> Should there be a popup or some kind of feedback?

only for the error case, see `try_or_popup` for examples.

> Should undo_last_commit() in utils.rs be more general function, for future use?

not sure what you mean?

**Comment by remique** [comment]
Unittests can probably be done better? Also, while working on this issue I have noticed that there are some commands with my keybinding (`u`) already assigned. To avoid potential conflicts, I changed `undo_commit` to `SHIFT-U` both in default and vim-keybindings (feel like it’s a bit more natural since it’s not the most used command in the world). Hopefully I didn’t forget about anything.

> not sure what you mean?

Yeah me neither, nevermind 😄

File: CHANGELOG.md [hunk 2]
```diff
@@ -15,6 +15,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 - indicator for longer commit message than displayed ([#773](https://github.com/extrawurst/gitui/issues/773))
 
 ![msg-len](assets/long-msg-indicator.gif)
+- undo-last-commit command under `[U]` key [[@remique](https://github.com/remique)] ([#758](https://github.com/extrawurst/gitui/issues/758))
```
Comment (extrawurst) [inline CHANGELOG.md:18]: this belongs under `Unreleased`

File: asyncgit/src/sync/utils.rs [hunk 3]
```diff
@@ -297,6 +297,54 @@ mod tests {
         Ok(())
     }
 
+    #[test]
+    fn test_undo_commit_empty_repo() -> Result<()> {
```
Comment (extrawurst) [inline asyncgit/src/sync/utils.rs:301]: no need for a `Result` since we `unwrap` everywhere (which is fine)

File: asyncgit/src/sync/utils.rs [hunk 4]
```diff
@@ -297,6 +297,54 @@ mod tests {
...
+    }
+
+    #[test]
+    fn test_undo_commit() -> Result<()> {
+        let (_td, repo) = repo_init().unwrap();
+        let root = repo.path().parent().unwrap();
+        let repo_path = root.as_os_str().to_str().unwrap();
+        let file_path = Path::new("file1.txt");
+
+        let status_count = |s: StatusType| -> usize {
```
Comment (extrawurst) [inline asyncgit/src/sync/utils.rs:319]: check out `get_statuses` we use for unittests already

File: asyncgit/src/sync/utils.rs [hunk 5]
```diff
@@ -297,6 +297,54 @@ mod tests {
...
+            get_status(repo_path, s).unwrap().len()
+        };
+
+        let full_path = &root.join(file_path);
```
Comment (extrawurst) [inline asyncgit/src/sync/utils.rs:323]: check out `write_commit_file` for convenience in unittests

File: asyncgit/src/sync/utils.rs [hunk 6]
```diff
@@ -297,6 +297,54 @@ mod tests {
...
+        File::create(full_path)
+            .unwrap()
+            .write_all(b"second commit content")
+            .unwrap();
+        stage_add_file(repo_path, second_file_path).unwrap();
+        commit(repo_path, "second commit").unwrap();
+
+        // Should be 1 file in staged
+        assert!(undo_last_commit(repo_path).is_ok());
+        assert_eq!(status_count(StatusType::Stage), 1);
```
Comment (extrawurst) [inline asyncgit/src/sync/utils.rs:343]: lets assert that the right file is now in status again

**Review by extrawurst — CHANGES_REQUESTED** [review]

**Comment by extrawurst** [comment]
good work @remique - almost there 👍

**Comment by remique** [comment]
Okay, so this definitely feels more natural of a test. Thank you so much for the help and hopefully everything is ready for the merge now.

**Comment by extrawurst** [comment]
@remique thanks ❤️

**Pull request merged**